implementations/rust/ockam/ockam_channel/src/secure_channel_worker.rs

```diff
@@ -0,0 +1,304 @@
+        &mut self,
+        ctx: &mut <Self as Worker>::Context,
+        payload: Vec<u8>,
+        is_first_initiator_msg: bool,
+    ) -> Result<()> {
+        if is_first_initiator_msg {
+            // First message from initiator goes to the channel listener
+            ctx.send_from_address(
+                self.remote_route.clone(),
+                CreateResponderChannelMessage::new(payload, self.first_responder_address.take()),
+                self.address_remote.clone(),
+            )
+            .await
+        } else {
+            // Other messages go to the channel worker itself
+            ctx.send_from_address(
+                self.remote_route.clone(),
+                payload,
+                self.address_remote.clone(),
+            )
+            .await
+        }
+    }
+}
+
+/// Key Exchange completed message
+#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
+pub struct KeyExchangeCompleted {
+    address: Address,
+    auth_hash: [u8; 32],
+}
+
+impl KeyExchangeCompleted {
+    /// Secure Channel address
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+    /// Authentication hash
+    pub fn auth_hash(&self) -> [u8; 32] {
+        self.auth_hash
+    }
+}
+
+#[async_trait]
+impl<V: SecureChannelVault, K: KeyExchanger + Send + 'static> Worker for SecureChannelWorker<V, K> {
+    type Message = Any;
+    type Context = Context;
+
+    async fn initialize(&mut self, ctx: &mut Self::Context) -> Result<()> {
+        if self.is_initiator {
+            if let Some(initiator) = self.key_exchanger.as_mut() {
+                let payload = initiator.process(&[])?;
+
+                self.send_key_exchange_payload(ctx, payload, true).await?;
+            } else {
+                return Err(SecureChannelError::InvalidInternalState.into());
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn handle_message(
+        &mut self,
+        ctx: &mut Self::Context,
+        msg: Routed<Self::Message>,
+    ) -> Result<()> {
+        let reply = msg.return_route().clone();
+        let mut onward_route = msg.onward_route();
+        let msg_addr = msg.msg_addr();
+        let transport_message = msg.into_transport_message();
+        let payload = transport_message.payload;
+
+        if msg_addr == self.address_local {
+            debug!("SecureChannel received Encrypt");
+
+            let _ = onward_route.step();
+
+            let msg = TransportMessage {
+                version: 1,
+                onward_route,
+                return_route: reply,
+                payload,
+            };
+            let payload = msg.encode()?;
+
+            let payload = {
+                let keys = Self::get_keys(&mut self.keys)?;
+
+                let nonce = keys.nonce;
+
+                if nonce == u16::max_value() {
+                    return Err(SecureChannelError::InvalidNonce.into());
+                }
+
+                keys.nonce += 1;
```

Comment:
I am not versed completely in the protocol, but is this increment by design or an implementation detail? Are there limitations to the transform applied here? i.e. MUST it be an increment, or could it be another random nonce?